### PR TITLE
refactor(common): move some over window utils to `common`

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

We had some handy utils and helper types in `over_window` module before, including `KeyWithSentinel` (renamed to `Sentinelled`), `EstimatedBTreeMap` and `DeltaBTreeMap`. This PR moves:

- `Sentinelled` to `common/types`;
- `EstimatedBTreeMap` to `common/estimate_size`;
- `DeltaBTreeMap` to `utils/delta_btree_map`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
Where can they be reused?

**Comment by stdrc** [comment]
> Where can they be reused?

Idk, but

- for `KeyWithSentinel`, I found it needed by another `over_window` submodule being implementing;
- for `EstimatedBTreeMap`, it's very common and has nothing related to over window, it's reasonable to move it to `common` following the convention of `EstimatedHashMap` and `EstimatedVecDeque`;
- for `DeltaBTreeMap`, we have a similar but with different purpose struct in meta, `BTreeMapTransaction`, which focuses on insert/delete/get operations while `DeltaBTreeMap` focuses on `get/cursor` operations. This indicates that this kind of helper types may be needed by other modules, so moving it to `common` can prevent potential re-invention.

**Comment by xxchan** [comment]
It would be better to be put in a separate crate `common/xxx` or `utils/xxx`

File: src/common/src/types/sentinel.rs [hunk 1]
```diff
@@ -13,17 +13,17 @@
...
 
 use enum_as_inner::EnumAsInner;
-use risingwave_common::estimate_size::EstimateSize;
-use risingwave_expr::window_function::StateKey;
+
+use crate::estimate_size::EstimateSize;
 
 #[derive(Debug, Clone, PartialEq, Eq, EnumAsInner)]
-pub(super) enum KeyWithSentinel<T> {
+pub enum Sentinelled<T> {
```
Comment (xxchan) [inline src/common/src/types/sentinel.rs:20]: Can you add some doc comments for it? It's not clear what it is.

**Comment by stdrc** [comment]
> It would be better to be put in a separate crate `common/xxx` or `utils/xxx`

Which one do you mean? This PR actually moves three different things to `common`: `Sentinelled`, `EstimatedBTreeMap` and `DeltaBTreeMap`. They logically belong to different `common/xxx`: `Sentinelled` is a type wrapper to help order things with **smallest** and **largest** which always rank first/last, so moved to `common/types`; `EstimatedBTreeMap` is a estimated-size version of `BTreeMap`, so moved to `common/estimated_size`; `DeltaBTreeMap` is a util to help iterate over a `BTreeMap` and some changes on it, so moved to `common/utils`.

File: src/common/src/types/sentinel.rs (hunk 1 above)
Comment (stdrc) [reply src/common/src/types/sentinel.rs:26]: Added.

**Comment by xxchan** [comment]
I mean I don't want to the `risingwave_common` crate bloats. (Ideally) each utils should live in it's own crate. Especially for sth that can be widely used by other ppl without any context, (and doesn't depend on other stuff).

I think the common crate is for things that are widely used in our system (and it's specialized for our system, e.g., types, arrays, chunks).

How do you think?>

**Comment by xxchan** [comment]
I'm emphasizing organizing code in "crates", not "modules" 🤪

**Comment by stdrc** [comment]
> I'm emphasizing organizing code in "crates", not "modules" 🤪

You're right. So we'd better separate many of the modules in `common` crate into different crates right?

**Comment by xxchan** [comment]
> So we'd better separate many of the modules in common crate into different crates right?

Yes, that's what in my head. Although I don't think it makes much difference for the `common` crate NOW (Currently it takes only a few secs to compile), I view it as a mindset problem. We can do the right things at the beginning.. Other opinions are also welcomed. 🤪

File: src/common/src/lib.rs [hunk 2]
```diff
@@ -44,6 +44,7 @@
 #![feature(negative_impls)]
 #![feature(bound_map)]
 #![feature(array_methods)]
+#![feature(btree_cursors)]
```
Comment (xxchan) [inline src/common/src/lib.rs:47]:
```suggestion
```

**Review by xxchan — APPROVED** [review]

File: src/common/src/lib.rs (hunk 2 above)
Comment (stdrc) [reply src/common/src/lib.rs:47]: This is still needed by `Sentinelled`

Comment (stdrc) [reply src/common/src/lib.rs:47]: This is still needed by `EstimatedBTreeMap`.

File: src/utils/delta_btree_map/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "delta_btree_map"
```
Comment (wenym1) [inline src/utils/delta_btree_map/Cargo.toml:2]: As our convention, the name shall be `risingwave_delta_btree_map`.

File: src/utils/delta_btree_map/src/lib.rs [hunk 4]
```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![feature(btree_cursors)]
```
Comment (wenym1) [inline src/utils/delta_btree_map/src/lib.rs:15]: The same as https://github.com/risingwavelabs/risingwave/pull/14204/files#r1436788805.

IIUC, methods related to the feature is not directly used in this crate, and therefore the feature is not required here. Only the `delta_btree_map` crate that uses the related feature needs to explicitly enable the feature.

**Review by wenym1 — APPROVED** [review]
Rest LGTM.

File: src/utils/delta_btree_map/src/lib.rs (hunk 4 above)
Comment (stdrc) [reply src/utils/delta_btree_map/src/lib.rs:15]: Actually it is directly used😇

https://github.com/risingwavelabs/risingwave/blob/d27d6c1eca115c60a6b31958e4a8ef6de5a5affe/src/common/src/estimate_size/collections/btreemap.rs#L95

**Pull request merged**